src/PresetReaction/PresetReaction.php

```diff
@@ -56,7 +56,11 @@ public function execute(EntityInterface $entity = NULL) {
     $action_ids = $config['actions'];
     foreach ($action_ids as $action_id) {
       $action = $this->actionStorage->load($action_id);
-      $action->execute([$entity]);
+      // Make sure that the action is appropriate:
+      // either system action or with same type as the entity type.
+      if ($action->getType() === 'system' || $entity->getEntityTypeId() === $action->getType()) {
+        $action->execute([$entity]);
+      }
```

Comment:
Checking the action type against the entity type prevents the action (meant for node (Islandora object)) to execute on a media.
Does including `empty($action->getType()) || $action->getType() === 'entity'` in the if statement solve the issues?
Or is this too crude and should the action itself determine if the action with the given entity can be performed? For example by using the access method of the Action class?
At the moment my understanding of derivers is limited; the code above is meant to make sure that an action is not called with the "wrong" entity type. If there is another way, I'm open to suggestions.

Reply:
I mean I think the _current_ onus is on the configured conditions to limit when an action is executed. 

It doesn't look like `access` is leveraged by any of the existing Islandora actions as of current. Given how the majority of the Islandora based actions just pop things on a queue to be processed by microservices at a `system` level most access restrictions provided on an individual user/entity won't be relevant. 

In normal Drupal land, limiting at the form configuration level would seem like the place to do it and would align with how core tends to operate. That is checking for action when retrieving actions whether it be on the entity query that gets executed on `loadMultiple` or filtered after the fact such as https://git.drupalcode.org/project/drupal/-/blob/9.5.3/core/modules/views/src/Plugin/views/field/BulkForm.php#L120-131 or https://git.drupalcode.org/project/search_api/-/blob/8.x-1.x/src/Plugin/views/field/SearchApiBulkForm.php#L42-49. With the former in mind that's not feasible given how these context reactions are configured to be setup without any entity context until runtime.

I think what you are proposing makes sense in that likely the `type` should be checked and used as a filtering factor here given the non-standard usage of `Context` and `Actions` within Islandora. If this change was to be introduced it'd likely need to be heavily documented as it would be breaking (whether it's introduced in a minor/major release).

Reply:
Isn't clear to me how targeting it there helps in this scenario? That is the conditions have already evaluated that an action should be fired and given the forced typification by the `MediaSourceService` it should be a media entity I think?

I know you mentioned it's tough to really have reproducible steps but perhaps we could connect in Slack to maybe identify the root of this? I have ran into a scenario before based upon underlying assumptions of context in general that it's meant to be operating on a single page. 

That is, take a scenario where you may have multiple contexts/actions triggering within a single thread with different entities being targeted. By its nature `Context` doesn't account for this scenario because in page rendering land it should never be accounted. To facilitate this we implemented our own `ContextRepository` and the actions that were exhibit behavior we moved over to using it (https://github.com/discoverygarden/dgi_actions/blob/1.x/src/ResettableContextRepository.php for reference). As we've unravelled in this thread it's a very hard thing to pinpoint down and reproduce in a more generic sense but perhaps this becomes a catalyst for introducing tests and a potential fix in the event something similar as it play here.

Reply:
Was discussed on https://github.com/Islandora/islandora-community/wiki/March-1%2C-2023 and the consensus seemed to be as follows:

- Split this issue into two where the `ContextManager` fixes can go in as they are straight bug fixes. A follow up issue could be made to track the other hardening around the execution of actions. In the latter's case it could be bundled to the next major version release such that its breaking nature could be documented and accounted for.

Other takeaways were as follows:
- @adam-vessey suggested a potential configuration checkbox toggle in the UI to enforce this type checking.
- @seth-shaw-asu also mentioned he had saw this on the starter site box in a reproducible form and he was going to attempt to lay out those steps which may help.

Reply:
I was able to reproduce the error messages on a starter site box (using the playbook).

1. I cloned https://github.com/seth-shaw-asu/migrate_apollo (technically, the precursor, but then I made a bunch of updates which have now been committed back to it)
2. I did the module's requisite copy of the data directory and enabling the module. (See the README.)
3. I then ran the migrations. That all worked fine, but...
4. I ran a _migrate rollback_ which then triggered the following errors:
```
vagrant@islandora8:/var/www/html/drupal$ drush --uri=https://localhost:8000 --userid=1 mr islandora8_metadata,islandora8_media,islandora8_file
 [notice] Now acting as user ID 1
 6/6 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%
 [notice] Rolled back 6 items - done with 'islandora8_media'
 6/6 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%
 [notice] Rolled back 6 items - done with 'islandora8_file'
 [error]  Error generating event: Bundle islandora_object does not exist 
 1/6 [▓▓▓▓░░░░░░░░░░░░░░░░░░░░░░░░]  16% [error]  Error generating event: Bundle islandora_object does not exist 
 2/6 [▓▓▓▓▓▓▓▓▓░░░░░░░░░░░░░░░░░░░]  33% [error]  Error generating event: Bundle islandora_object does not exist 
 3/6 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓░░░░░░░░░░░░░░]  50% [error]  Error generating event: Bundle islandora_object does not exist 
 4/6 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓░░░░░░░░░░]  66% [error]  Error generating event: Bundle islandora_object does not exist 
 5/6 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓░░░░░]  83% [error]  Error generating event: Bundle islandora_object does not exist 
 6/6 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%
 [notice] Rolled back 6 items - done with 'islandora8_metadata'
 [notice] Switching back from user 1.
 [error]  Message: Error generating event: Bundle islandora_object does not exist
```
Applying the PR caused those "Error generating event" messages to not appear.

Reply:
Thanks @seth-shaw-asu will look again!

Reply:
So, figured this out.

It's is indeed the same issue we saw at dgi in terms of contexts not being reset between evaluations. That is, previously evaluated contexts within the same thread are retained when additional conditions are evaluated. 

@lucasvanschaik at this point can you remove the action filtering so the `Context` fixes can be merged and I'll open a new issue / PR for the follow-up once I can figure out how best to package all this up?

Reply:
Contained in https://github.com/Islandora/islandora/pull/932.